File: emulator/teclaemu.java
/**
 *
 * @author Rishabh
 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.HashMap;
import java.util.Map;
import javax.bluetooth.*;
import javax.microedition.io.*;

public class Emulator {
    
    private final UUID uuid = new UUID("27012f0c68af4fbf8dbe6bbaf7aa432a", false);    
    private final String name = "TeclaShield";
    private final String url = "btspp://localhost:" + uuid
                                + ";name=" + name
                                + ";authenticate=false;encrypt=false;";
    private LocalDevice localDevice = null;
    private StreamConnectionNotifier server = null;
    private StreamConnection conn = null;
    private InputStream din = null;
    private OutputStream dout = null;
    private Map<String,Integer> dict;
    private BufferedReader reader = null;
    private String helpString = "";
    private PingThread pinging;
    private boolean auto_release = true;

    public Emulator(){
        dict = new HashMap<String,Integer>();
        dict.put("W",0x3E);
        dict.put("S", 0x3D);
        dict.put("A", 0x3B);
        dict.put("D", 0x37);
        dict.put("R", 0x3F);
        pinging = new PingThread();
        auto_release = true;
        reader = new BufferedReader(new InputStreamReader(System.in));
        helpString += 
                "**********HELP**********\n"
                +"w/W : ECU1 pressed\n"
                +"s/S : ECU2 pressed\n"
                +"a/A : ECU3 pressed\n"
                +"d/D : ECU4 pressed\n"
                +"1 : Switch Port 1 pressed\n"
                +"2 : Switch port 2 pressed\n"
                +"h/H : View possible commands\n"
                +"r/R : Generate realease switch event\n"
                + "t/T : Toggle auto switch release event\n"
                + "q/Q : Quit\n"
                + "\n\n#####Auto switch release mode is a mode in which switch release events are inserted after every switch event\n"
                + "Default set to true.\n"
                + "Can be turned on or off by command t/T\n"
                +"************************";
        System.out.println(helpString);
    }
    
    private boolean discover(){
        try {
            localDevice = LocalDevice.getLocalDevice();
            localDevice.setDiscoverable(DiscoveryAgent.GIAC);
            server = (StreamConnectionNotifier) Connector.open(url);
            System.out.println("Waiting for connection...");
            conn = server.acceptAndOpen();
            System.out.println("Device connected...");
            din = conn.openInputStream();
            dout = conn.openOutputStream();
        } catch (Exception ex) {
            System.out.println("An error occured while connecting to the device");
            return false;
        }
        return true;
    }
    
    private boolean sendKeyEvent(int keyEvent){
        try {
            synchronized(dout){
                System.out.println("writing : " + keyEvent);
                dout.write(keyEvent);
                if(auto_release){
                    try {
                        Thread.sleep(200);
                    } catch (InterruptedException ex) {}
                    dout.write(dict.get("R"));
                }
            }
        } catch (IOException ex) {
            System.out.println("An error occured while sending to the connected device");
            return false;
        }
        return true;
    }
    
    class PingThread extends Thread{

        private boolean end = false;

        @Override
        public void run() {
            end = false;
            while(!end){
                try {
                    try {
                        this.sleep(100);
                    } catch (InterruptedException ex) {}
                    int read = din.read();
                    if(read != -1)
                    synchronized(dout){
                        dout.write(read);
                    }
                } catch (IOException ex) {
                    System.out.println("An error occured while pinging the connected device, disconnecting");
                    disconnect();
                    System.exit(0);
                }
            }
        }
        
        public void end(){
            end = true;
        }

    };
    
    private void disconnect(){
        try {
            din.close();
        } catch (Exception ex) {}
        try {
            dout.close();
        } catch (Exception ex) {}
        try {
            server.close();
        } catch (Exception ex) {}
    }

    private void startEmulator(){
        if(!this.discover()){
            System.out.println("Closing the emulator");
            System.exit(0);
        }
        pinging.start();
        this.listenKeys();
    }
    
    private void listenKeys(){
        System.out.println("Enter the keystroke (h for help): ");
        while(true){
            char ans = 'h';
            try {
                String g = reader.readLine();
                ans = (g.equals(""))?' ':g.charAt(0);
            } catch (IOException ex) {
                System.out.println("An internal error occured, exiting");
                pinging.stop();
                disconnect();
                System.exit(0);
            }
            boolean sendKeyEvent = true;
            if((ans == 'w' || ans == 'W') && dict.containsKey("W")){
                sendKeyEvent = this.sendKeyEvent(dict.get("W"));
            }
            else if((ans == 's' || ans == 'S') && dict.containsKey("S")){
                sendKeyEvent = this.sendKeyEvent(dict.get("S"));
            }
            else if((ans == 'a' || ans == 'A') && dict.containsKey("A")){
                sendKeyEvent = this.sendKeyEvent(dict.get("A"));
            }
            else if((ans == 'd' || ans =='D') && dict.containsKey("D")){
                sendKeyEvent = this.sendKeyEvent(dict.get("D"));
            }
            else if((ans == 'r' || ans =='R') && dict.containsKey("R")){
                sendKeyEvent = this.sendKeyEvent(dict.get("R"));
            }
            else if(ans == '1' && dict.containsKey("1")){
                sendKeyEvent = this.sendKeyEvent(dict.get("1"));
            }
            else if(ans == '2' && dict.containsKey("2")){
                sendKeyEvent = this.sendKeyEvent(dict.get("2"));
            }
            else if(ans == 'h'|| ans =='H'){
                System.out.println(helpString);
            }
            else if(ans == 't' || ans =='T'){
                auto_release = !auto_release;
                if(auto_release)
                    System.out.println("Auto release is now on");
                else
                    System.out.println("Auto release is now off");
            }
            else if(ans == 'q' || ans =='Q'){
                pinging.stop();
                this.disconnect();
                System.exit(0);
            }
            if(!sendKeyEvent){
                System.out.println("An unknown error occured while sending this keystroke to the connected device, exiting");
                pinging.stop();
                this.disconnect();
                System.exit(0);                
            }
        }
    }
    
    public static void main(String[] args){
        Emulator emulator = new Emulator();
        emulator.startEmulator();
    }
}

File: emulator/teclaemu.py
# -*- coding: utf-8 -*-
import time
from bluetooth import *
import threading

#
# Tecla Emulator v1.2
# written by Akhil Rao
# Extended by Ankit Daftery
#

## {{{ http://code.activestate.com/recipes/134892/ (r2)
## Modified to support non-blocking I/O
class _Getch:
	"""Gets a single character from standard input. Does not echo to the
screen."""
	def __init__(self):
		try:
			self.impl = _GetchWindows()
		except ImportError:
			self.impl = _GetchUnix()
	
	def __call__(self): return self.impl()


class _GetchUnix:
	def __init__(self):
		import tty, sys
	
	def __call__(self):
		import sys, tty, termios
		from select import select		# Using select for non-blocking I/o
		fd = sys.stdin.fileno()
		old_settings = termios.tcgetattr(fd)
		try:
			tty.setraw(sys.stdin.fileno())
			[i, o, e] = select([sys.stdin.fileno()], [], [], 5)		# Timeout in 5 seconds if no input
			if i: ch=sys.stdin.read(1)
			else: ch='.'		# Dummy character to avoid KeyError in sendEvent
		finally:
			termios.tcsetattr(fd, termios.TCSADRAIN, old_settings)
		return ch


class _GetchWindows:
	def __init__(self):
		import msvcrt
	
	def __call__(self):
		import msvcrt
		return msvcrt.getch()

getch = _Getch()
## end of http://code.activestate.com/recipes/134892/ }}}


class TeclaThread(threading.Thread):
	"""
	Threaded Class to handle sockets to the TeclaShields
	"""
	def __init__(self,client_socket,address):
		self.client_socket = client_socket
		self.address = address
		self.exitflag=False
		self.c="."								#Null character initialised to avoid transmitting 
		threading.Thread.__init__(self)
	
	def disconnect(self):
		"""
		Close the socket
		"""
		self.client_socket.close()
		self.exitflag=True
	
	def sendEvent(self,event):
		"""
		Transmit a Shield keyevent to TeclaAccess
		"""
		try:
			self.client_socket.send(chr(keyvalue[event]))
		except BluetoothError,err:
			print "Bluetooth Error in TeclaThread",err
	
	def listenkeys(self):
		"""
		Function to generate keyevents based on user input
		"""
		global flag		#Global is necessary to allow thread synchronisation
		self.auto_release = True;
		self.exitflag=False;
		while not self.exitflag :
			print ("Your switch action w/s/a/d/1/2/r?");
			self.c= getch()
			self.c = self.c.upper();
			if(self.c in "WSAD12QHRT"):
				#Valid key pressed
				if(self.c in "HTQ"):
					#Command key pressed
					if self.c == "H": #Help
						print helpstring
					if self.c == "T": #Help
						self.auto_release = not self.auto_release;
						if self.auto_release:
							print "Auto-release ON"
						else:
							print "Auto-release OFF"
					if self.c == "Q": #Hel
						self.disconnect()
						flag=True
						print keymessage[self.c]
				
				if(self.c in "WSAD12R"):
					#Switch key pressed
					self.sendEvent(self.c)
					print keymessage[self.c];
					if self.auto_release:
						time.sleep(0.1) #Connection seems to have long latency
						self.sendEvent('R')
			else:
				if('.' in self.c):
					print "No input received\n"
				#Invalid key pressed
				else:
					print "Invalid value %s"%(self.c)
	
	def run(self):
		"""
		Overriden run function
		"""
		print 'Accepted connection from ', self.address
		self.listenkeys()

#Keyboard 
#ECU1 -> Highlight next 
#ECU2 -> Highlight prev
#ECU3 -> Cancel
#ECU4 -> select highlighted
#
#SP1 -> Select highlighted
#SP2 -> Cancel

#keyvalue is the dictionary used for holding character input from keyboard and corresponding value of the byte to be sent....
keyvalue={
	"R":0x3F,		#Released all switches
	"W":0x3E,		#ECU1 pressed
	"S":0x3D,		#ECU2 pressed
	"A":0x3B,		#ECU3 pressed
	"D":0x37,		#ECU4 pressed
	"1":0x2F,		#Switch Port 1 pressed
	"2":0x1F,		#Switch Port 2 pressed
	"Q":0x88,		#Quit the emulator
	"H":0x77,		#Show commands list
	"T":0x99,		#Toggle release event mode
	}

#keymessage dictionary to notify users of action performed on keypress
keymessage= {
	"R":"All Switches Released ",
	"W":"ECU1 pressed",
	"S":"ECU2 pressed",
	"A":"ECU3 pressed",
	"D":"ECU4 pressed",
	"1":"Switch Port 1 pressed",
	"2":"Switch Port 2 pressed",
	"Q":"Quitting the emulator"
	}

#Help menu
helpstring = "\nw/W => Generate Event on Joystick 1"
helpstring=helpstring + "\ns/S => Generate Event on Joystick 2 "
helpstring=helpstring + "\na/A => Generate Event on Joystick 3 "
helpstring=helpstring + "\nd/D => Generate Event on Joystick 4 "
helpstring=helpstring + "\n1 => Generate Event on Switch Port 1"
helpstring=helpstring + "\n2 => Generate Event on Switch Port 2"
helpstring=helpstring + "\nh/H => view possible commands"
helpstring=helpstring + "\nr/R => Generate release switch event"
helpstring=helpstring + "\nt/T => Toggle auto switch release event"
helpstring=helpstring + "\nq/Q => Quit "
helpstring=helpstring + "\n\n\n#####Auto switch release mode is a mode in which switch release events are inserted after every switch event \nDefault set to true.\nCan be turned on or off by command t/T"


#Initial notifss
version = "v0.2"
print 'Tecla - Emulator ' , version ;
print 'press h/H for list of possible commands'
print 'press q/Q to quit'
print 'Auto switch release mode set to True'

#setup the bluetooth RFComm socket i.e bind and start listening
server_socket = BluetoothSocket( RFCOMM )
server_socket.bind(('', PORT_ANY))
server_socket.listen(1)

#uuid can be used if bluetooth client side program is written to minimize discovery and filtering time
uuid = "00001101-0000-1000-8000-00805F9B34FB"
advertise_service(
	server_socket,
	"SPP",
	uuid,
	service_classes=[SERIAL_PORT_CLASS],
	profiles=[SERIAL_PORT_PROFILE]
	)

#Global variable to check for quit
flag = False

while (not flag):
	print 'Waiting for a device ........'
	client_socket, address = server_socket.accept()
	t = TeclaThread(client_socket,address)
	t.start()
	while t.isAlive():
		time.sleep(0.1)
		try:
			data_in=t.client_socket.recv(128)
			if(len(data_in) > 0):
				t.client_socket.send(data_in)
		except BluetoothError:
			t.disconnect()
			t.join()
			break

#Clean up
server_socket.close()
client_socket.close()
print "Disconnected"
